interactor/src/sovereign_forge/sovereign_forge_interactor_main.rs

```diff
@@ -125,79 +148,202 @@ impl SovereignForgeInteract {
         fee: Option<FeeStruct<StaticApi>>,
     ) {
-        self.deploy_template_contracts().await;
-        self.deploy_sovereign_forge(&deploy_cost).await;
+        let initial_caller = self.bridge_owner_shard_0.clone();
+
+        let sovereign_forge_address = self
+            .deploy_sovereign_forge(initial_caller.clone(), &BigUint::from(DEPLOY_COST))
+            .await;
 
-        let sov_forge_address = self.state.current_sovereign_forge_sc_address().clone();
-        let chain_config_address = self.state.current_chain_config_sc_address().clone();
-        let mvx_esdt_safe_address = self.state.current_mvx_esdt_safe_contract_address().clone();
-        let fee_market_address = self.state.current_fee_market_address().clone();
-        let header_verifier_address = self.state.current_header_verifier_address().clone();
+        for shard_id in 0..NUMBER_OF_SHARDS {
+            let caller = self.get_sovereign_owner_for_shard(shard_id);
+            let template_contracts = self
+                .deploy_template_contracts(caller.clone(), EsdtSafeType::MvxEsdtSafe)
+                .await;
+
+            let (
+                chain_config_address,
+                mvx_esdt_safe_address,
+                fee_market_address,
+                header_verifier_address,
+            ) = match template_contracts.as_slice() {
+                [a, b, c, d] => (a.clone(), b.clone(), c.clone(), d.clone()),
+                _ => panic!(
+                    "Expected 4 deployed contract addresses, got {}",
+                    template_contracts.len()
+                ),
+            };
+
+            self.finish_init_setup_phase_for_one_shard(
+                shard_id,
+                initial_caller.clone(),
+                sovereign_forge_address.clone(),
+                TemplateAddresses {
+                    chain_config_address: chain_config_address.clone(),
+                    header_verifier_address: header_verifier_address.clone(),
+                    esdt_safe_address: mvx_esdt_safe_address.clone(),
+                    fee_market_address: fee_market_address.clone(),
+                },
+            )
+            .await;
+            println!("Finished setup phase for shard {shard_id}");
+        }
+
+        for shard in 0..NUMBER_OF_SHARDS {
+            self.deploy_on_one_shard(
+                shard,
+                deploy_cost.clone(),
+                optional_esdt_safe_config.clone(),
+                optional_sov_config.clone(),
+                fee.clone(),
+            )
+            .await;
+        }
+    }
+
+    pub async fn finish_init_setup_phase_for_one_shard(
+        &mut self,
+        shard_id: u32,
+        initial_caller: Address,
+        sovereign_forge_address: Address,
+        template_addresses: TemplateAddresses,
+    ) {
+        let caller = self.get_bridge_owner_for_shard(shard_id);
+        let preferred_chain_id = PREFERRED_CHAIN_IDS[shard_id as usize].to_string();
```

Comment:
Why do you need `preferred_chain_id`? You need to get address based on shard-id, not chain-id